File: leetcode/HouseRobber_198.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int rob(vector<int>& nums) {
        vector<int> take(nums.size());
        vector<int> leave(nums.size());

        take[0] = nums[0];
        for(int i = 1; i < nums.size(); ++i) {
            take[i] = leave[i - 1] + nums[i];
            leave[i] = max(leave[i - 1], take[i - 1]);
        }

        return max(take[take.size() - 1], leave[leave.size() - 1]);
    }
};

File: leetcode/HouseRobber_198.go
package leetcode

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func rob(nums []int) int {
	take := 0
	leave := 0

	take = nums[0]
	for i := range nums {
		if i == 0 {
			continue
		}

		temp := take
		take = leave + nums[i]
		leave = max(leave, temp)
	}
	return max(leave, take)
}
